include/runtime/collections/RBTree.h

```diff
@@ -140,14 +138,37 @@ class RBTree {
   T const &root_key() const {
     assert(!empty());
     const InternalNode *r = static_cast<const InternalNode *>(root_.get());
-    return r->key_;
+    return r->data_.first;
   }
 
   // Return the value stored in the root Node of this tree.
   V const &root_val() const {
     assert(!empty());
     const InternalNode *r = static_cast<const InternalNode *>(root_.get());
-    return r->val_;
+    return r->data_.second;
+  }
+
+  // Return a const reference to the data (key-value pair) stored in the root
+  // Node of this tree.
+  std::pair<T, V> const &root_data() const {
+    assert(!empty());
+    const InternalNode *r = static_cast<const InternalNode *>(root_.get());
+    return r->data_;
+  }
+
+  /* WARNING: The following method returns a non-const reference in order to  *
+   * enable altering this data structure (and potentially ones implemented on *
+   * top of it, namely RangeMap) in place. The data structure is intended to  *
+   * be immutable: the exposed functionality should not alter the data        *
+   * structure's contents in place. Only request a non-const reference if you *
+   * in fact need to edit the data structure in place for a specific reason,  *
+   * e.g. garbage collection.                                                 */
```

Comment:
Perhaps worth just expanding slightly on this point - I presume the restriction is that you can't use this to modify the contents in a way that affects the tree's internal invariants? e.g. the tree ordering isn't looking at the GC bits

Reply:
Addressed.